numpy/f2py/src/fortranobject.c

```diff
@@ -690,32 +709,120 @@ static int swap_arrays(PyArrayObject* obj1, PyArrayObject* obj2) {
-                                PyObject *obj) {
+static int
+get_elsize(PyObject *obj) {
+  /*
+    get_elsize determines array itemsize from a Python object.  Returns
+    elsize if succesful, -1 otherwise.
+
+    Supported types of the input are: numpy.ndarray, bytes, str, tuple,
+    list.
+  */
+
+  if (PyArray_Check(obj)) {
+    return PyArray_DESCR((PyArrayObject *)obj)->elsize;
+  } else if (PyBytes_Check(obj)) {
+    return PyBytes_GET_SIZE(obj);
+  } else if (PyUnicode_Check(obj)) {
+    return PyUnicode_GET_LENGTH(obj);
+  } else if (PySequence_Check(obj)) {
+    PyObject* fast = PySequence_Fast(obj, "f2py:fortranobject.c:get_elsize");
+    if (fast != NULL) {
+      Py_ssize_t i, n = PySequence_Fast_GET_SIZE(fast);
+      int sz, elsize = 0;
+      for (i=0; i<n; i++) {
+        sz = get_elsize(PySequence_Fast_GET_ITEM(fast, i) /* borrowed */);
+        if (sz > elsize) {
+          elsize = sz;
+        }
+      }
+      Py_DECREF(fast);
+      return elsize;
+    }
+  }
+  return -1;
+}
+
+extern PyArrayObject *
+ndarray_from_pyobj(const int type_num,
+                   const int elsize_,
+                   npy_intp *dims,
+                   const int rank,
+                   const int intent,
+                   PyObject *obj,
+                   const char *errmess) {
     /*
+     * Return an array with given element type and shape from a Python
+     * object while taking into account the usage intent of the array.
+     *
+     * - element type is defined by type_num and elsize
+     * - shape is defined by dims and rank
+     *
+     * ndarray_from_pyobj is used to convert Python object arguments
+     * to numpy ndarrays with given type and shape that data is passed
+     * to interfaced Fortran or C functions.
+     *
+     * errmess (if not NULL), contains a prefix of an error message
+     * for an exception to be triggered within this function.
+     *
+     * Negative elsize value means that elsize is to be determined
+     * from the Python object in runtime.
+     *
+     * Note on strings
+     * ---------------
+     *
+     * String type (type_num == NPY_STRING) does not have fixed
+     * element size and, by default, the type object sets it to
+     * 0. Therefore, for string types, one has to use elsize
+     * argument. For other types, elsize value is ignored.
+     *
+     * NumPy defines the type of a fixed-width string as
+     * dtype('S<width>'). In addition, there is also dtype('c'), that
+     * appears as dtype('S1') (these have the same type_num value),
+     * but is actually different (.char attribute is either 'S' or
+     * 'c', respecitely).
+     *
+     * In Fortran, character arrays and strings are different
+     * concepts.  The relation between Fortran types, NumPy dtypes,
+     * and type_num-elsize pairs, is defined as follows:
+     *
+     * character*5 foo     | dtype('S5')  | elsize=5, shape=()
+     * character(5) foo    | dtype('S1')  | elsize=1, shape=(5)
```

Comment:
The two Fortran calls should be equivalent (though * should come with a warning), so I'm not sure I understand why they resolve to different cases here.

Reply:
No, strictly speaking, these are not equivalent: `character*5` is a character string with a given length while `character(5)` is an array of characters with a given dimension.
`*` is valid Fortran 77, so no warning is necessary.

Reply:
I thought CHARACTER type specifications took only KIND and LEN as defined in `7.4.4.2 Character type specifier` from the [F2018 Interpretation Document](https://gitlab.com/lfortran/lfortran/uploads/973db2bfcc64cd2a397f823c32718b39/18-007r1.pdf).

```fortran
character(5) :: a ! LEN=5 from type-param-value
character :: a(5) ! LEN=1 by default, a is an array
character*5 :: a ! LEN=5 by * char-length
```

![image](https://user-images.githubusercontent.com/4336207/132218449-a3775878-7e6b-4811-a02a-b02412d993a6.png)

From `9.5.3 Array elements and array sections`, I was under the impression that `characterTHING` is always a character specifier and the array rules only come into effect for `var(DIM)` declarations from.

Reply:
I don't have Fortran 77 std at hand but see for instance:
- https://www.ibiblio.org/pub/languages/fortran/ch2-13.html
- https://docs.oracle.com/cd/E19957-01/805-4939/6j4m0vn7r/index.html
that illustrate the difference between character strings, character arrays, and character string arrays.

That said, I see your point re `character(5)` and `character*5` equivalence and the dimension specifications. Let me re-check the implementation...

Reply:
It is surprisingly difficult to hunt down the F77 manual, since the WG5 documents only go back to S8.99

However I found [an HTML archive](https://wg5-fortran.org/ARCHIVE/Fortran77.html) on the WG5 page. The Character definition does not seem to clearly mention LEN or KIND.  However `5.2.3  Size_of_an_Array` mentions:

> If the actual  argument  is  a  character  array
                 name, character array element name, or character
                 array  element  substring  name  and  begins  at
                 character  storage  unit  t_  of an array with c_
                 character storage units, then the  size  of  the
                 dummy  array  is INT((c_ + 1 - t_) / ln__), where ln__
                 is the length of an element of the dummy array.

Which seems to be something not mentioned in the F90 standard.

I think interpreting `CHARACTER*(6+3) C ` as described by the ORCALE page to be an array given by `CHARACTER C*(6+3)` is non-standard.

Reply:
Ah I found a more canonical reference and example. `1.2.2 Character Variables and Arrays ` of the `ANSI-77_FORTRAN_Information_Document` (Internet Archive edition) mentions:

![image](https://user-images.githubusercontent.com/4336207/132260400-338fee6e-89cc-482e-8098-7a19f958e416.png)

This is on the basis of the fact that the form of a character type-statement was simpler.

![image](https://user-images.githubusercontent.com/4336207/132260431-cb9c56d0-499e-40fd-816e-083f40518ebc.png)

In later standards the array distinction becomes clearer. To sum things up then:

```fortran
GAMMA(10)*8
```

Is problematic since F77 seems to accept this as a `character` array with `length 8`; while F90 onwards this would have `length 10` and be undefined from my reading. Practically:

```fortran
program main
  implicit none
  character*5 :: fooi(5)
  character GAMMA(10)*8
  fooi = "12345"
  GAMMA = "1h2d5r6"
  print*, fooi
  print*, GAMMA
end program
```

Under `gfortran` this interprets both of these the same for some reason; that is `GAMMA` is taken to be of length 8 and is an array of 10 (it gets repeated). This is actually clearly a bug since by the F2018 interpretation () is interpreted first as LEN.

Reply:
A practical way forward would then be the following:
- Interpret `a [(d)]*len` to be F77 always
- Otherwise follow F90